cmd/snap-update-ns/change.go

```diff
@@ -591,106 +586,124 @@ func neededChangesOld(currentProfile, desiredProfile *osutil.MountProfile) []*Ch
 		}
 	}
 
-	// Mount desired entries not reused.
-	for i := range desired {
-		if !reuse[desired[i].Dir] {
-			changes = append(changes, &Change{Action: Mount, Entry: desired[i]})
+	var desiredNotReused []osutil.MountEntry
+	for _, entry := range desired {
+		if !reuse[entry.Dir] {
+			desiredNotReused = append(desiredNotReused, entry)
 		}
 	}
 
-	return changes
-}
-
-// neededChangesNew is the real implementation of NeededChanges
-// This function is used when RobustMountNamespaceUpdate is enabled.
-func neededChangesNew(currentProfile, desiredProfile *osutil.MountProfile) []*Change {
-	// Copy both profiles as we will want to mutate them.
-	current := make([]osutil.MountEntry, len(currentProfile.Entries))
-	copy(current, currentProfile.Entries)
-	desired := make([]osutil.MountEntry, len(desiredProfile.Entries))
-	copy(desired, desiredProfile.Entries)
+	// Mount desired entries not reused, ordering by the mimic directories they
+	// need created
+	// We proceeds in three steps:
+	// 1. Perform the mounts for the "overname" entries
+	// 2. Perform the mounts for the entries which need a mimic
+	// 3. Perform all the remaining desired mounts
 
-	// Clean the directory part of both profiles. This is done so that we can
-	// easily test if a given directory is a subdirectory with
-	// strings.HasPrefix coupled with an extra slash character.
-	for i := range current {
-		current[i].Dir = filepath.Clean(current[i].Dir)
-	}
-	for i := range desired {
-		desired[i].Dir = filepath.Clean(desired[i].Dir)
+	var newDesiredEntries []osutil.MountEntry
+	// Indexed by mount point path.
+	addedDesiredEntries := make(map[string]bool)
+	ensureDesiredEntry := func(entry osutil.MountEntry) {
```

Comment:
we tend to call this kind of helpers addDesiredEntry even they are idempotent. You can add a comment about it being done "once"

Reply:
Done.